### PR TITLE
store: support fine-grained store changes notifications

### LINKED ISSUE #134: store: more granular state changes notifications
The current _Store_ listener notifies that _something_ changed in the store, so the user can update their views, re-run queries, etc reacting accordingly.

This is quite limiting since not having extra information about what changed might force the user to run expensive queries again which may not be necessary.

We slightly divised a possible solution for this in #126:
> > (sander) I _think_ eventually we'll want to enable the store listener to push the latest state, w/o requiring an explicitly query. This would allow the client to just listen for all entities changing.
> 
> (ignacio) Yeah, totally. The current notification is just a _signal_ that _something_ changed, which falls pretty short. I like the idea of giving different levels of granularity for notifications, like three options:
> 
>     * Something changed in Store (what exists now). (Good for multi-model aggregates or views)
> 
>     * Something changed in a Model. (Good for aggregates/views of single models)
> 
>     * Something changed in a particular instance of a Model. (Good for track particular things)

### PR DESCRIPTION
**Pull request by jsign**
Fixes #134 

This PR adds support for a new way of listening changes in `Store`. The old way that simply returned a `chan struct{}` (and good luck!) is gone.

I migrated all uses of old API to the new one that is more general.

Main idea:
The `Store` has a `Listen(..)` API. 
This API receives whatever the number of optional filters to configure what changes you want to hear in flexible ways.
Examples:
- Listen to all changes: `l, err := store.Listen()` (call without args)
- Listen to all changes of some model: `l, err := store.Listen(ListenOption{Model: "Person"})`
- More complex example: listen to Saves in Model2 *or* Saves in Model1 with EntityID "id-i2" *or* deletes of Model1 with id-i1 *or* all deletes from Model2:
```
opts := []ListenOption{ 
			ListenOption{Model: "Model2", Type: ListenSave},
			ListenOption{Model: "Model1", Type: ListenSave, ID: "id-i2"},
			ListenOption{Model: "Model1", Type: ListenDelete, ID: "id-i1"},
			ListenOption{Model: "Model2", Type: ListenDelete},
}
l, err := store.Listen(opts...)
```

As you can see, you can concatenate filters with `Model`, `Type`, `ID` (any of them optional), to allow most possible combinations of wanted things. e.g: if you don't specify a `Type` for the `ListenOption` would mean all types (save, delete, create))

That's for the "filtering" part.

This `l` returned, has `l.Channel()`, which is: `chan Action`:
```
type Action struct {
	Model string
	Type  ActionType
	ID    core.EntityID
}
```
Saying it differently, now what's received in the channel isn't a `struct{}`, is an `Action` (filtered by whatever you chose) describing what exactly changed in Store: in which model, which entityID, what happened (Create, Save, or Delete).

See the new tests to get an extra feeling of how should work.

Depends on [merging a PR from `go-textile-core`](https://github.com/textileio/go-textile-core/pull/32 ) which is ready for _review_. After merged, I'll change the `go.mod` of this PR and will be ready for official review (CirceCI tests will pass when this is done).

Added some comments to help review; ask anything if isn't clear enough!

### CODE REVIEW

**Comment by asutula** [comment]
Just reading your description, this looks perfect!

File: eventstore/listeners.go [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+	"fmt"
+	"sync"
+
+	"github.com/textileio/go-textile-core/broadcast"
+	core "github.com/textileio/go-textile-core/store"
+)
+
+// Listen returns a StoreListener which notifies about actions applying the
+// defined filters. The Store *won't* wait for slow receivers, so if the
+// channel is full, the action will be dropped.
```
[the remark this answers is not in the document]
Comment (sanderpick) [reply eventstore/listeners.go]: Makes sense to me.

**Comment by jsign** [comment]
Updated `go.mod`, when tests pass, I'll merge.

**Comment by jsign** [comment]
A particular test fails `TestThreads/AddPeer` which was a flaky one for quite some time. I see it already failing in other PRs. Maybe it is a CircleCI dependant thing since locally all tests pass.

Opened #153 to track this.

**Pull request merged**